src/protocol/argument_parser.rs

```diff
@@ -0,0 +1,118 @@
+}
+
+impl Parser {
+    fn parse_error<A>(&self, message: &str) -> R<A> {
+        Err(format!("{} ({:?})", message, self.original))?
+    }
+
+    fn skip_spaces(&mut self) {
+        loop {
+            match self.input.pop_front() {
+                Some(' ') => {}
+                Some(char) => {
+                    self.input.push_front(char);
+                    break;
+                }
+                None => break,
+            }
+        }
+    }
+
+    fn skip_char(&mut self, expected: char, message: &str) -> R<()> {
+        let next = self.input.pop_front();
+        if next != Some(expected) {
+            self.parse_error(message)?;
+        }
+        Ok(())
+    }
+
+    fn peek_chars(&mut self, expected: Vec<Option<char>>, message: &str) -> R<()> {
+        let peeked = self.input.get(0);
+        if !expected.contains(&peeked.cloned()) {
+            self.parse_error(message)?;
+        }
+        Ok(())
+    }
+
+    fn parse_char(&mut self, excluded: &[char]) -> R<Option<char>> {
+        Ok(match self.input.pop_front() {
+            None => None,
+            Some('\\') => Some(match self.input.pop_front() {
+                None => self.parse_error("a backslash must be followed by a character")?,
+                Some('"') => '"',
+                Some('n') => '\n',
+                Some(' ') => ' ',
+                Some('\\') => '\\',
+                Some(char) => self.parse_error(&format!("unknown escaped character {}", char))?,
+            }),
+            Some(char) => {
+                if excluded.contains(&char) {
+                    self.input.push_front(char);
+                    None
+                } else {
+                    Some(char)
+                }
+            }
+        })
+    }
+
+    fn collect_chars_until(&mut self, excluded: &[char]) -> R<String> {
+        let mut result = "".to_string();
+        while let Some(char) = self.parse_char(excluded)? {
+            result.push(char);
+        }
+        Ok(result)
+    }
+
+    fn parse_word(&mut self) -> R<Option<String>> {
+        self.skip_spaces();
+        Ok(match self.input.get(0) {
+            None => None,
+            Some('"') => {
+                self.skip_char('"', "shouldn't happen")?;
+                let word = self.collect_chars_until(&['"'])?;
+                self.skip_char('"', "unmatched quotes")?;
+                self.peek_chars(
+                    vec![Some(' '), None],
+                    "closing quotes must be followed by a space",
+                )?;
+                Some(word)
+            }
+            Some(_) => {
+                let result = self.collect_chars_until(&[' ', '"'])?;
+                self.peek_chars(
+                    vec![Some(' '), None],
+                    "opening quotes must be preceeded by a space",
+                )?;
+                Some(result)
+            }
+        })
+    }
+
+    pub fn parse_arguments(arguments: &str) -> R<Vec<String>> {
```

Comment:
The only public function now is `parse_arguments`, which returns a `Result<Vec<String>, _>` :tada: 

This is mostly a pleasant abstraction over the `Iterator` implementation below, so that we don't have to use cumbersome turbofish calls to `collect`